docs/plugins.md

```diff
@@ -1 +1,105 @@
+
+2. Customizations - while Kn intends to provide a useful and complete CLI for the Knative APIs, customers of Knative and Kn may have additional commands that they want to add to complete or customize the resulting user experience. For instance, installation and updates commands. Or for instance a `doctor` command to help heal the Knative installation. These commands may be specific to the targeted cloud or Kube cluster. A plugin command could solve these idiosyncratic needs.
+
+Other motivation might exist. And when they surface, from our users' actual usage of the plugin in Kn, we will list them here.
+
+## Use-cases
+
+Since we have divided the motivation for plugins into two general categories, let's explore some known use cases for each category.
+
+### Experimentations
+
+1. `doctor` plugin command - to perform som diagnostics on the Knative cluster and perhaps list actions (or perform them automatically) to heal the cluster.
+2. `pods` plugin command - to list the pods for a currently running Knative service.
+3. `ssh` plugin command - to automatically ssh into a running pod hosting a Knative service.
+4. `monitor` plugin command - to aggregate and stream the logs and thus help monitor a running Knative service.
+5. `event-flow` plugin command - to combine a series of events and perform some trigger on a Knative service.
+
+### Customizations
+
+1. `install` plugin command - a cloud-specific install/uninstall command. For instance, installing and uninstalling Knative on the IBM Kubernetes Service (IKS). This needs to make use of IKS-specific commands and setup and credentials. The same could be said for other Kubernetes services in the market.
+2. `update` plugin command - a command to update the Knative installation.
+3. `migrate` plugin command - a command to migrate services from a proprietary or other OSS FaaS platform to Knative.
+
+## Architecture
+
+The current proposal is to follow the Kubectl plugin architecture. The following is a brief summary of this architecture. 
+
+* Any executable on the terminal's PATH that starts with `kn-`, e.g., `/usr/local/bin/kn-doctor`
+* All user parameters and flags are passed to the plugin executable as is
+* All environment variables are passed to the plugin executable as is
+* A warning is displayed to the user if the command tries to shadow another plugin in a different part of the PATH
+* A error is displayed to the user if the command tries to shadow an existing Kn command, e.g., a plugin named `kn-service-list`
+* For plugins exposing commands with groups and subcommands, then the name needs to also include dashes. So a plugin `kn pods list` would be named `kn-pods-list` 
```

Comment:
Wouldn't it just be easier to let the command itself dispatch on subcommands (which come in as additional arguments ?). This would leave the structure lean to only one level (like for kubectl plugins). I mean a `kn-pods` could check the arguments for a "list" and the does its work.

Reply:
It's implemented as in `kubectl`. I tried to reuse the same *exact* code. Perhaps my interpretation in the doc is incorrect?